vault/resource_database_secret_backend_role.go

```diff
@@ -20,121 +24,176 @@ var (
 				Type:        schema.TypeString,
 				Required:    true,
 				ForceNew:    true,
 				Description: "Database connection to use for this role.",
 			},
-			"default_ttl": {
+			consts.FieldDefaultTTL: {
 				Type:        schema.TypeInt,
 				Optional:    true,
 				Description: "Default TTL for leases associated with this role, in seconds.",
 			},
-			"max_ttl": {
+			consts.FieldMaxTTL: {
 				Type:        schema.TypeInt,
 				Optional:    true,
 				Description: "Maximum TTL for leases associated with this role, in seconds.",
 			},
-			"creation_statements": {
+			consts.FieldCreationStatements: {
 				Type:        schema.TypeList,
 				Required:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Database statements to execute to create and configure a user.",
 			},
-			"revocation_statements": {
+			consts.FieldRevocationStatements: {
 				Type:        schema.TypeList,
 				Optional:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Database statements to execute to revoke a user.",
 			},
-			"rollback_statements": {
+			consts.FieldRollbackStatements: {
 				Type:        schema.TypeList,
 				Optional:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Database statements to execute to rollback a create operation in the event of an error.",
 			},
-			"renew_statements": {
+			consts.FieldRenewStatements: {
 				Type:        schema.TypeList,
 				Optional:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Description: "Database statements to execute to renew a user.",
 			},
+			consts.FieldCredentialType: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Specifies the type of credential that will be generated for the role.",
+			},
+			consts.FieldCredentialConfig: {
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Description: "Specifies the configuration for the given credential_type.",
+			},
+			consts.FieldKeyBits: {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  2048,
+				Description: "Number of bits to use for the generated keys. Options include: 2048 " +
+					"(default), 3072, 4096; with key_type=ec, allowed values are: 224, 256 (default), " +
+					"384, 521; ignored with key_type=ed25519.",
+			},
+			consts.FieldSignatureBits: {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  256,
+				Description: "The number of bits to use in the signature algorithm. Options include: 256 " +
+					"(default), 384, 512.",
+			},
+			consts.FieldFormat: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The output format of the generated private key credential.",
+			},
+
+			consts.FieldPasswordPolicy: {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "The policy used for password generation. Only used when credential type " +
+					"is 'password'.",
+			},
```

Comment:
Should any of these new fields utilize `RequiredWith` to prevent users from providing only a subset of values? (For example is "key_type" required when setting any of the above values?)

Reply:
I think we'll let Vault handle the extra bits of validation for the different subsets for `credential_config` here. In some cases it can be beneficial to break earlier, but from our testing we found that providing wrong subsets results in meaningful errors from Vault anyway, so adding extra validation was something we decided to skip. If we find that from feedback in the future that users would like to break early before they ever communicate with Vault, we can always add this bit in